gateway/mw_oauth2_auth.go

```diff
@@ -143,34 +152,35 @@ func generateCacheKey(config apidef.UpstreamOAuth, apiId string) string {
 	return hex.EncodeToString(hash.Sum(nil))
 }
 
-func (cache *upstreamOAuthCache) getToken(r *http.Request, OAuthSpec *UpstreamOAuth) (string, error) {
+func (cache *upstreamOAuthClientCredentialsCache) getToken(r *http.Request, OAuthSpec *UpstreamOAuth) (string, error) {
 	cacheKey := generateCacheKey(OAuthSpec.Spec.UpstreamAuth.OAuth, OAuthSpec.Spec.APIID)
 
-	token, err := cache.retryGetKeyAndLock(cacheKey)
+	tokenString, err := retryGetKeyAndLock(cacheKey, &cache.RedisCluster)
 	if err != nil {
 		return "", err
 	}
 
-	if token != "" {
-		decryptedToken := decrypt(getPaddedSecret(OAuthSpec.Gw), token)
+	if tokenString != "" {
+		decryptedToken := decrypt(getPaddedSecret(OAuthSpec.Gw), tokenString)
 		return decryptedToken, nil
 	}
 
-	token, err = cache.obtainToken(r.Context(), OAuthSpec)
+	token, err := cache.obtainToken(r.Context(), OAuthSpec)
 	if err != nil {
 		return "", err
 	}
 
-	encryptedToken := encrypt(getPaddedSecret(OAuthSpec.Gw), token)
+	encryptedToken := encrypt(getPaddedSecret(OAuthSpec.Gw), token.AccessToken)
 
-	if err := cache.setTokenInCache(cacheKey, encryptedToken); err != nil {
+	ttl := time.Duration(token.Expiry.Sub(time.Now()).Seconds()) * time.Second
+	if err := setTokenInCache(cacheKey, encryptedToken, ttl, &cache.RedisCluster); err != nil {
 		return "", err
 	}
 
-	return token, nil
+	return token.AccessToken, nil
 }
 
-func (cache *upstreamOAuthCache) retryGetKeyAndLock(cacheKey string) (string, error) {
+func retryGetKeyAndLock(cacheKey string, cache *storage.RedisCluster) (string, error) {
```

Comment:
is the timeout for lock (5 seconds) to be a hardcoded value?

Reply:
I thought 5 to be a sensible value as any OAuth provider should not go over this duration. If you have a different suggestion please leave it here